### PR TITLE
Document SQL SEQSCAN

### LINKED ISSUE #3281: SQL: document a new SEQSCAN keyword for the SELECT scan queries
**Product:** Tarantool
**Since:** 2.11
**SME:** @ ImeevMA
**Root document:** https://www.tarantool.io/en/doc/latest/reference/reference_sql/sql_statements_and_clauses/#sql-select
https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_space/_session_settings/

Also, check these links:
* https://www.tarantool.io/en/doc/latest/how-to/sql/sql_tutorial/
* https://www.tarantool.io/en/doc/latest/reference/reference_sql/sql-features/
* https://www.tarantool.io/en/doc/latest/how-to/sql/sql_beginners_guide/

Keep in mind that the default value for sql_seq_scan session setting is also described in the compat module docs (since v.3.0): https://www.tarantool.io/en/doc/latest/reference/reference_lua/compat/sql_seq_scan_default/

# Details
Now scanning SELECT will not run and will throw an error if the new
SEQSCAN keyword is not used for scanned spaces. This change only affects
SELECT and does not affect UPDATE and DELETE. A SELECT is recognized as
a scanning SELECT if `EXPLAIN QUERY PLAN SELECT ...` indicates that the
SELECT `scans` rather than `searches`.

For example, if we have spaces created with these queries:
```
CREATE TABLE t(i INT PRIMARY KEY, a INT);
CREATE TABLE s(i INT PRIMARY KEY, a INT);
```

Then these queries will throw an error:
```
SELECT * FROM t;
SELECT * FROM t WHERE a > 1;
SELECT * FROM t WHERE i + 1 = 5;
SELECT * FROM t, s;
SELECT * FROM t JOIN s;
```

And these will not:
```
SELECT * FROM t WHERE i > 1;
SELECT * FROM SEQSCAN t;
SELECT * FROM SEQSCAN t WHERE i + 1 = 5;
SELECT * FROM SEQSCAN t, SEQSCAN s;
SELECT * FROM SEQSCAN t JOIN SEQSCAN s;
```

Scanning can be allowed or disallowed by default. To do this, a new
session setting is introduced: `sql_seq_scan`. The default value for
setting is `true`, i.e. scanning is allowed. When set to `false`, the
scanning SELECTs will throw a `scanning is not allowed` error.
Requested by @ImeevMA in https://github.com/tarantool/tarantool/commit/77648827326ad268ec0ffbcd620c2371b65ef2b4.

# Definition of done

- [x] SQL reference contains and describes the new keyword `SEQSCAN` in the [FROM clause](https://www.tarantool.io/en/doc/latest/reference/reference_sql/sql_statements_and_clauses/#from-clause) section.
- [x] [Session settings list](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_space/_session_settings/) contains the new setting `sql_seq_scan`.
- [x] All SQL `SELECT` examples across the docs (mostly in [SQL how-tos](https://www.tarantool.io/en/doc/latest/how-to/sql/)) contain the SEQSCAN keyword if the do perform the sequential scanning.
- [x] S[QL user guide](https://www.tarantool.io/en/doc/latest/reference/reference_sql/sql_user_guide/) mentions the `SEQSCAN` keyword.

### PR DESCRIPTION
**Pull request by p7nov**
Resolves #3281 

- Added the SEQSCAN description to the FROM clause reference: https://docs.d.tarantool.io/en/doc/gh-3281-sql-seqscan/reference/reference_sql/sql_statements_and_clauses/#from-clause
- Added SEQSCAN to the list of reserved words in the SQL user guide: https://docs.d.tarantool.io/en/doc/gh-3281-sql-seqscan/reference/reference_sql/sql_user_guide/#reserved-words
- Added the SEQSCAN section to the SQL tutorial and updated all SELECT examples across this tutorial with SEQSCAN (where it is required): https://docs.d.tarantool.io/en/doc/gh-3281-sql-seqscan/how-to/sql/sql_tutorial/#the-seqscan-keyword
- Added a note about allowing scan queries to the SQL beginners' guide: https://docs.d.tarantool.io/en/doc/gh-3281-sql-seqscan/how-to/sql/sql_beginners_guide/#selecting-with-where
- Added `sql_seq_scan` session setting to the session settings reference: https://docs.d.tarantool.io/en/doc/gh-3281-sql-seqscan/reference/reference_lua/box_space/_session_settings/

### CODE REVIEW

File: doc/how-to/sql/sql_tutorial.rst [hunk 1]
```diff
@@ -117,6 +117,54 @@ Then try to put another row:
 This ``INSERT`` fails because of a primary-key violation: the row with the primary
 key ``1, 'AB'`` already exists.
 
+The SEQSCAN keyword
+~~~~~~~~~~~~~~~~~~~
+
+In Tarantool, SQL queries that perform sequential scans (that is, go through all
```
Comment (ImeevMA) [inline doc/how-to/sql/sql_tutorial.rst]: I think it's better to mention that this only applies to `SELECT` queries and subqueries.

File: doc/how-to/sql/sql_tutorial.rst [hunk 2]
```diff
@@ -117,6 +117,54 @@ Then try to put another row:
...
+
+    SELECT * FROM table2
+
+To execute a scan query, put the ``SEQSCAN`` keyword before the table name:
+
+.. code-block:: sql
+
+    SELECT * FROM SEQSCAN table2
+
+If a query uses indexed fields in filters, it can still be a scan query.
```
Comment (ImeevMA) [inline doc/how-to/sql/sql_tutorial.rst]: In our terminology, `scan queries` are queries that do not use indexes. This means that if a `SELECT` query uses indexes, then that query is not a `scan query`, so it can be executed without `SEQSCAN` keyword.

File: doc/how-to/sql/sql_tutorial.rst [hunk 3]
```diff
@@ -184,8 +235,8 @@ The rows which have the same values for ``column2`` are grouped and are aggregat
 
 .. code-block:: sql
 
-    SELECT column2, SUM(column4), COUNT(column4), AVG(column4)
-    FROM table2
+    SELECT SEQSCAN column2, SUM(column4), COUNT(column4), AVG(column4)
```
Comment (ImeevMA) [inline doc/how-to/sql/sql_tutorial.rst]: I think this `SEQSCAN` is unnecessary. We use `SEQSCAN` for tables, not for columns.

File: doc/how-to/sql/sql_tutorial.rst [hunk 4]
```diff
@@ -117,6 +117,54 @@ Then try to put another row:
...
+The SEQSCAN keyword
+~~~~~~~~~~~~~~~~~~~
+
+In Tarantool, SQL queries that perform sequential scans (that is, go through all
+the table instead of using indexes) are prohibited by default. For example, this
+query leads to the error ``Scanning is not allowed for 'table2'``:
+
+.. code-block:: sql
+
+    SELECT * FROM table2
```
Comment (ImeevMA) [inline doc/how-to/sql/sql_tutorial.rst]: I think `;` should be added since we set `;` as delimiter.

File: doc/how-to/sql/sql_tutorial.rst [hunk 5]
```diff
@@ -117,6 +117,54 @@ Then try to put another row:
...
+
+To execute a scan query, put the ``SEQSCAN`` keyword before the table name:
+
+.. code-block:: sql
+
+    SELECT * FROM SEQSCAN table2
```
Comment (ImeevMA) [inline doc/how-to/sql/sql_tutorial.rst]: I think `;` is missing.

File: doc/how-to/sql/sql_tutorial.rst [hunk 6]
```diff
@@ -117,6 +117,54 @@ Then try to put another row:
...
+The result is:
+
+*   The first query returns rows:
+
+    .. code-block:: tarantoolsession
+
+        - [1, 'AB', 'AB', 10.5]
+        - [1, 'CD', '  ', 10005]
+
+*   The second query fails with the error ``Scanning is not allowed for 'TABLE2'``.
```
Comment (ImeevMA) [inline doc/how-to/sql/sql_tutorial.rst:153]: It might be better to mention that the second query is a scan query, since such an expression cannot be executed using index.

It might be worth mentioning somewhere which expressions can be executed using indexes. However, this can be quite a large section since we also need to describe how multi-field indexes work..

File: doc/how-to/sql/sql_tutorial.rst [hunk 7]
```diff
@@ -152,7 +203,7 @@ Retrieve some of what you inserted:
 
 .. code-block:: sql
 
-    SELECT column1, column2, column1 * column4 FROM table2 WHERE column2
+    SELECT column1, column2, column1 * column4 FROM SEQSCAN table2 WHERE column2
     LIKE 'A%';
     SELECT column1, column2, column3, column4 FROM table2
```
Comment (ImeevMA) [inline doc/how-to/sql/sql_tutorial.rst]: The `SEQSCAN` keyword is missing.

File: doc/how-to/sql/sql_tutorial.rst [hunk 8]
```diff
@@ -516,7 +567,7 @@ with arithmetic on a number column and ordering by a string column.
     INSERT INTO t6 VALUES (+1234567890, 'GD', 1e30);
     INSERT INTO t6 VALUES (10, 'FADEW?', 0.000001);
     INSERT INTO t6 VALUES (5, 'ABCDEFG', NULL);
-    SELECT column1 + 1, column2, column4 * 2 FROM t6 ORDER BY column2;
+    SELECT column1 + 1, column2, column4 * 2 FROM SEQSCAN t6 ORDER BY column2;
```
Comment (ImeevMA) [inline doc/how-to/sql/sql_tutorial.rst:571]: In `CREATE VIEW v3 AS SELECT SUBSTR(column2,1,2), column4 FROM t6 WHERE column4 >= 0;` we should also add `SEQSCAN` keyword. Otherwise `SELECT * FROM v3;` fails.

Same for `WITH cte AS (SELECT SUBSTR(column2,1,2), column4 FROM t6 WHERE column4 >= 0) SELECT * FROM cte;`.

Same for `SELECT * FROM "_space";`, `SELECT * FROM "_index` and `SELECT * FROM "_trigger";`.

File: doc/reference/reference_sql/sql_statements_and_clauses.rst [hunk 9]
```diff
@@ -1889,18 +1889,35 @@ Parentheses are allowed, and ``[[AS] correlation-name]`` is allowed.
...
 
+The ``SEQSCAN`` keyword (since :doc:`2.11 </release/2.11.0>`) marks the queries that
+perform sequential scans during the execution. It happens if the query can't use indexes,
+and goes through all the table rows one by one, sometimes causing a heavy load.
+Such queries are called *scan queries*. If a scan query doesn't have the
+``SEQSCAN`` keyword, Tarantool raises an error. ``SEQSCAN`` must precede all
+names of the tables that the query scans.
+
+To find out if a query performs a sequential scan, use ``EXPLAIN QUERY PLAN``.
+For scan queries, the result contains ``SCAN_TABLE``.
```
Comment (ImeevMA) [inline doc/reference/reference_sql/sql_statements_and_clauses.rst]: I think it is better to write `SCAN TABLE ...` or `SCAN TABLE table_name` instead of `SCAN_TABLE`. For example:
```
tarantool> box.execute([[EXPLAIN QUERY PLAN SELECT * FROM "_space", "_index" WHERE "_space"."id" == 512;]])
---
- metadata:
  - name: selectid
    type: integer
  - name: order
    type: integer
  - name: from
    type: integer
  - name: detail
    type: text
  rows:
  - [0, 0, 0, 'SEARCH TABLE _space USING PRIMARY KEY (id=?) (~1 row)']
  - [0, 1, 1, 'SCAN TABLE _index (~1048576 rows)']
...
```

**Review by ImeevMA — COMMENTED** [review]
Hi! Thank you for the patch. I have some comments, please take a look at them.

File: doc/how-to/sql/sql_tutorial.rst (hunk 8 above)
Comment (p7nov) [reply doc/how-to/sql/sql_tutorial.rst:571]: Thanks, fixed these examples.

File: doc/how-to/sql/sql_tutorial.rst (hunk 6 above)
Comment (p7nov) [reply doc/how-to/sql/sql_tutorial.rst:153]: Rewrote the example description.

File: doc/reference/reference_sql/sql_statements_and_clauses.rst [hunk 10]
```diff
@@ -1889,18 +1889,35 @@ Parentheses are allowed, and ``[[AS] correlation-name]`` is allowed.
...
 
+The ``SEQSCAN`` keyword (since :doc:`2.11 </release/2.11.0>`) marks the queries that
+perform sequential scans during the execution. It happens if the query can't use indexes,
+and goes through all the table rows one by one, sometimes causing a heavy load.
+Such queries are called *scan queries*. If a scan query doesn't have the
+``SEQSCAN`` keyword, Tarantool raises an error. ``SEQSCAN`` must precede all
+names of the tables that the query scans.
+
+To find out if a query performs a sequential scan, use ``EXPLAIN QUERY PLAN``.
+For scan queries, the result contains ``SCAN_TABLE tabe_name``.
```
Comment (ImeevMA) [inline doc/reference/reference_sql/sql_statements_and_clauses.rst]: It is actually `SCAN TABLE table_name` instead of `SCAN_TABLE tabe_name`. See example in the comment above.

**Review by ImeevMA — APPROVED** [review]
Thank you for the fixes! I have only one comment.

File: doc/reference/reference_sql/sql_statements_and_clauses.rst [hunk 11]
```diff
@@ -1889,18 +1889,35 @@ Parentheses are allowed, and ``[[AS] correlation-name]`` is allowed.
...
+names of the tables that the query scans.
+
+To find out if a query performs a sequential scan, use ``EXPLAIN QUERY PLAN``.
+For scan queries, the result contains ``SCAN TABLE table_name``.
+
+..  note::
+
+    For backward compatibility, the scan queries without the ``SEQSCAN`` keyword
+    are allowed in Tarantool 2.11. The errors on scan queries will be the default
+    behavior starting from 3.0. You can change the default behavior of scan queries
```
Comment (andreyaksenov) [inline doc/reference/reference_sql/sql_statements_and_clauses.rst:1906]: > will be the default behavior starting from 3.0

I'd rephrase this to the present tense - 3.0 (albeit an alpha version) already published and the behavior is already changed. Looks like we can eliminate a future promise here. Additionally, we won't update this note to the present tense when the 3.0 release version is actually published.

**Review by andreyaksenov — APPROVED** [review]

File: doc/reference/reference_sql/sql_statements_and_clauses.rst (hunk 11 above)
Comment (p7nov) [reply doc/reference/reference_sql/sql_statements_and_clauses.rst:1906]: Thanks, fixed.
